json-decode.go

```diff
@@ -12,65 +12,52 @@ import (
 //----------------------------------------
 // cdc.decodeReflectJSON
 
-// CONTRACT: rv is addressable as per https://golang.org/pkg/reflect/#Value.CanAddr
-func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) error {
-	// By this point, we MUST be dealing with an addressable value.
+// CONTRACT: rv.CanAddr() is true.
+func (cdc *Codec) decodeReflectJSON(bz []byte, info *TypeInfo, rv reflect.Value, opts FieldOptions) (err error) {
 	if !rv.CanAddr() {
-		panic("value is not addressable")
+		panic("rv not addressable")
 	}
-
-	// SANITY CHECK
 	if info.Type.Kind() == reflect.Interface && rv.Kind() == reflect.Ptr {
 		panic("should not happen")
 	}
 
-	// No need for diambiguation, decode as is.
+	// No need for disambiguation, decode as is.
 	if !info.Registered {
-		return cdc._decodeReflectJSON(bz, info, rv, opts)
+		err = cdc._decodeReflectJSON(bz, info, rv, opts)
+		return
 	}
 
-	// Otherwise disambiguation time
-	disfix, restBlob, err := parseDisfixAndData(bz)
+	// Otherwise, disambiguation time.
+	disfix, bz, err := decodeDisfixJSON(bz)
 	if err != nil {
-		return err
+		return
 	}
-	info, err = cdc.getTypeInfoFromDisfix_rlock(disfix)
+	cinfo, err := cdc.getTypeInfoFromDisfix_rlock(disfix)
 	if err != nil {
-		return err
+		return
 	}
 
-	bz = restBlob
 	// And we need to construct the concrete type
 	// that'll then be set into the interface field.
-	crv, _ := constructConcreteType(info)
-	if err := cdc._decodeReflectJSON(bz, info, crv, opts); err != nil {
-		return err
+	var crv, _ = constructConcreteType(cinfo)
+	if err = cdc._decodeReflectJSON(bz, info, crv, opts); err != nil {
+		rv.Set(crv) // Helps with debugging
+		return
 	}
-	rv.Set(crv)
-	return nil
-}
 
-// constructConcreteType creates the concrete value as
-// well as the corresponding settable value for it.
-func constructConcreteType(cinfo *TypeInfo) (crv, rvSet reflect.Value) {
-	// Construct new concrete type.
-	// NOTE: rv.Set() should succeed because it was validated
-	// already during Register[Interface/Concrete].
-	if cinfo.PointerPreferred {
-		cPtrRv := reflect.New(cinfo.Type)
-		crv = cPtrRv.Elem()
-	} else {
-		crv = reflect.New(cinfo.Type).Elem()
-		rvSet = crv
-	}
-	return crv, rvSet
+	// We need to set here, for when !PointerPreferred and the type
```

Comment:
We don't use PointerPreferred in our JSON codec unless I missed a memo

Reply:
Hmm we should, for when we decode an interface value.

Reply:
constructConcreteType does use it, whether JSON or binary.